### ISSUE TITLE
Cursor position is lost when resizing the window if not on the bottom line

### ISSUE THREAD

**Issue by daanx** [microsoft/terminal#10848]
### Windows Terminal version (or Windows build number)

10.0.19042.1110 (windows terminal 1.9.1942.0)

### Steps to reproduce

If the cursor position is not on the last line (of the input), making the screen more narrow will push the cursor down to the last line when lines wrap and the original position is lost; on a resize to make it wider again it stays on the last line.

This affects utilities like "readline" when editing multiple lines of input. Hope it can be fixed as it makes multiline editing not work at all.  

To reproduce, run the following batch file:
```bat
@echo off
echo Resize the window to make it more narrow and wider again.
echo The cursor will lose its position and move down.
echo.
echo The cursor is first positioned at the start of this [33mlong[0m line.
set /p "ANSWER=Press <enter> to stop ..[1F"
echo.
echo Thank you.
```
(note: I think the paste lost it but there should be an `ESC` character in front of the `[` characters to make the escape sequence function; the `ESC[1F` moves the cursor one up so it is no longer on the last line.)

### Expected Behavior

The cursor should always stay at the same point in the input regardless  of a window resize. The "old" cmd terminal is correct and the cursor stays at the same point..

### Actual Behavior

![cursor-bug-cmd](https://user-images.githubusercontent.com/2171100/127802782-ae3a8a8a-5d30-4dbd-92ac-e41ef3726d4a.png)

Here you see the cursor in Windows Terminal (I made it red) move down a line on the resize and lose its position.

**Comment by zadjii-msft** [comment]
Thanks for the helpful repro case! I'm gonna link this up to #5800, because this might be a repro for a heisenbug I've been hunting

**Comment by zadjii-msft** [comment]
Okay my notes: 
this repros when the line is close enough to the bottom line of the terminal window, that the `Press <enter>` line will get placed on exactly the bottom line. When the `long` line wraps (which will create a blank line between the `long` line and the `<enter>` line), then the cursor gets left at the start of the `<enter>` line, instead of the `long` one. 


Curiously, in `conhost`, if the buffer size == viewport size, then the cursor gets left at the _end_ of the `long` line. It moves down one row, to the created blank one, then back up to the end of the `long` line when you de-flow the long line.

**Comment by zadjii-msft** [comment]
This is because we'll end up circling the buffer during the resize, _after_ we found the cursor position. 

Say the window has 20 rows. That puts the cursor on row 18. When we resize down, and reflow the buffer, we'll look for the old cursor. We'll find it at (0,18) in the old buffer, when we're writing to (0,19) in the new buffer. This is because `long` line 1 already reflowed, creating an extra row in the buffer. 

So we'll stash (0,19) as the expected cursor position in the new buffer. We'll reflow this whole line into the buffer, which will cause us to circle! The place the cursor should _really_ be is line 17, not 19. 

I'm putting this in the same bucket as #10130. I think this is gonna just go away with #8000, though we're probably _"out of the frying pan, into the fire"_ with that one

**Comment by daxian-dbw** [comment]
I observed the same. Basically, the cursor is not guaranteed to point to the same character that it was originally pointing at when the window resizing happens. This happens often enough, and makes it difficult for PowerShell to calculate the position of the initial prompt after a resizing.

![Animation](https://user-images.githubusercontent.com/127450/143963370-2d777e14-13f1-4dfb-8183-2d5f29cafef2.gif)

**Comment by lhecker** [comment]
Same here... I can't seem to reproduce this issue in 1.17/1.18. Did we fix this?

**Comment by lhecker** [comment]
(Sorry for the other comment I posted here - I meant to post it in #10868.)

I believe this issue as described is fixed, but #14448 gave me an idea for reproducing a similar, probably related issue.

Using this .bat file:
```bat
@echo off
set /p "ANSWER=Could you share maybe a screenshot/? What shell are you using, and/or what program is doing the outputting? I bet this is #3088 just based on the description, but I can't be sure."
```

and by resizing the window randomly it is possible to place the cursor in the wrong position. From what I can tell, the issue is due to a desync between conhost and Windows Terminal, because Windows Terminal resizes its buffer before conhost does. This most obvious when actively printing text of course, but it also happens when regularly resizing. I believe my PR #15701 exacerbates this issue, because the reflow is so much faster giving more time for a desync to happen.

But as I said, I believe that the two issues you filed, as they're described, are fixed, even if reflow in general is still buggy.
